src/daft-functions/src/temporal/truncate.rs

```diff
@@ -0,0 +1,78 @@
+use common_error::{DaftError, DaftResult};
+use daft_core::prelude::*;
+use daft_dsl::{
+    functions::{ScalarFunction, ScalarUDF},
+    ExprRef,
+};
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub struct Truncate {
+    interval: String,
+}
+
+#[typetag::serde]
+impl ScalarUDF for Truncate {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn name(&self) -> &'static str {
+        stringify!($fn_name)
+    }
+
+    fn to_field(&self, inputs: &[ExprRef], schema: &Schema) -> DaftResult<Field> {
+        match inputs {
+            [input, relative_to] => match (input.to_field(schema), relative_to.to_field(schema)) {
+                (Ok(input_field), Ok(relative_to_field))
+                    if input_field.dtype.is_temporal()
+                        && (relative_to_field.dtype.is_temporal()
+                            || relative_to_field.dtype.is_null()) =>
+                {
+                    Ok(Field::new(input_field.name, input_field.dtype))
+                }
+                (Ok(input_field), Ok(relative_to_field)) => Err(DaftError::TypeError(format!(
+                    "Expected temporal input args, got {} and {}",
+                    input_field.dtype, relative_to_field.dtype
+                ))),
+                (Err(e), _) | (_, Err(e)) => Err(e),
```

Comment:
Super minor nit, but maybe we could use the `?` operator to avoid an unwieldy match statement. Something like:

```rs
let input_field = input_field.to_field(schema)?;
let relative_to = relative_to.to_field(schema)?;
if input.dtype.is_temporal() && .. {
  Ok(Field::new(..))
} else {
  Err(DaftError::TypeError(..))
}
```